ts/src/profiler.ts

```diff
@@ -13,47 +13,292 @@
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * Stackdriver Profiler, this problem will be logged at the debug level.
+   * If there is a problem polling Stackdriver Profiler for instructions
+   * on the type of profile created, this problem will be logged. If the problem
+   * indicates one definitely will not be able to profile, an error will be
+   * thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent.'));
+    return this.pollProfilerService();
+  }
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   */
+  async pollProfilerService(): Promise<void> {
+    const startCreateMillis = Date.now();
+    const prof = await this.createProfile();
+    await this.profileAndUpload(prof);
+    const endCreateMillis = Date.now();
+
+    // Schedule the next profile.
+    setImmediate(this.pollProfilerService.bind(this)).unref();
+  }
+
+  /**
+   * Talks to Stackdriver Profiler server, which hangs until server indicates
+   * job should be profiled.
+   *
+   * If any problem is encountered, the problem will be logged and
+   * createProfile() will be retried.
+   *
+   * TODO: implement backoff and retry when error encountered. createProfile()
+   * should be retried at time response indicates this request should be retried
+   * or with exponential backoff (up to one hour) if the response does not
+   * indicate when to retry this request. Once this is implemented, an error
+   * will be thrown only if the error indicates one definitely should not
+   * retry createProfile.
+   *
+   * Public to allow for testing.
+   */
+  async createProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: {
+        projectId: this.config.projectId,
+        target: this.config.serviceContext.service,
+        labels: {zone: this.config.zone, instance: this.config.instance},
+      },
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    let requestError: Error|undefined = undefined;
+    let retryRequest = true;
+    try {
+      const [body, response] = await this.request(options);
+      if (isErrorResponseCode(response.statusCode)) {
+        retryRequest = isRetriableResponseCode(response.statusCode);
+        requestError =
+            new Error('Error creating profile: ' + response.statusMessage);
+        this.logger.debug(requestError);
+      } else {
+        return body;
+      }
+    } catch (err) {
+      retryRequest =
+          isRetriableError(err) || isRetriableResponseCode(err.statusCode);
+      requestError = new Error('Error creating profile: ' + err.toString());
+      this.logger.debug(requestError);
```

Comment:
Why logged as a debug level and not an error?

Reply:
Switched to logging at error level -- that seems to be what is done for profiler.go